File: tools/jayd-library/firmware_reader_self_check.py
#!/usr/bin/env python3
"""Build converter fixtures and exercise the firmware metadata reader on the host."""

from __future__ import annotations

import binascii
import json
import shutil
import struct
import subprocess
import sys
import tempfile
from pathlib import Path


HERE = Path(__file__).resolve().parent
ROOT = HERE.parent.parent
CONVERTER = HERE / "jayd_library.py"
READER = ROOT / "src" / "Metadata" / "JaydMetadata.cpp"
CHECK = HERE / "firmware_reader_self_check.cpp"
HOST_INCLUDE = HERE / "host"
HEADER = struct.Struct("<8sHHHHIIIQQI16s")
SECTION = struct.Struct("<4sHHIIQ")


def repaired(data: bytearray) -> bytes:
    struct.pack_into("<I", data, 44, 0)
    struct.pack_into("<I", data, 44, binascii.crc32(data) & 0xFFFFFFFF)
    return bytes(data)


def section_index(data: bytes, kind: bytes) -> int:
    _, _, _, _, entry_size, count, *_ = HEADER.unpack_from(data)
    for index in range(count):
        if data[HEADER.size + index * entry_size:HEADER.size + index * entry_size + 4] == kind:
            return index
    raise AssertionError(f"missing section {kind!r}")


def fixtures(directory: Path) -> None:
    source = directory / "engine.json"
    source.write_text(json.dumps({
        "schema": "jayd-engine-interchange-1",
        "producer": "firmware-reader-self-check",
        "tracks": [{
            "source_id": "stable-source-1",
            "path": "safe/song.aac",
            "sample_rate": 44100,
            "duration_frames": 110250,
            "bpm": "128.125",
            "key": "8A",
            "rating": 4,
            "cues": [
                {"kind": "cue", "position_frames": 22050, "label": "Cue"},
                {"kind": "loop", "position_frames": 44100, "length_frames": 22050, "label": "Loop"},
            ],
            "beatgrid": [{
                "position_frames": 0,
                "bpm": "128.125",
                "beat_number": 1,
                "confidence": 9000,
            }],
            "phrases": [{
                "position_frames": 0,
                "kind": "intro",
                "confidence": 8000,
            }],
        }],
    }), encoding="utf-8")
    subprocess.run([
        sys.executable, str(CONVERTER), "engine-json", str(source),
        "-o", str(directory / "valid.jydm"),
    ], check=True, stdout=subprocess.DEVNULL)
    valid = (directory / "valid.jydm").read_bytes()

    corrupt = bytearray(valid)
    corrupt[-1] ^= 1
    (directory / "corrupt-crc.jydm").write_bytes(corrupt)
    (directory / "truncated.jydm").write_bytes(valid[:-1])

    oversized = bytearray(valid)
    struct.pack_into("<I", oversized, 20, 4097)
    (directory / "oversized-count.jydm").write_bytes(repaired(oversized))

    overflow = bytearray(valid)
    meta = section_index(overflow, b"META")
    struct.pack_into("<Q", overflow, HEADER.size + meta * SECTION.size + 16, 0xFFFFFFFFFFFFFFFC)
    (directory / "offset-overflow.jydm").write_bytes(repaired(overflow))

    overlap = bytearray(valid)
    cues = section_index(overlap, b"CUES")
    tracks = section_index(overlap, b"TRAK")
    track_offset = struct.unpack_from("<Q", overlap, HEADER.size + tracks * SECTION.size + 16)[0]
    struct.pack_into("<Q", overlap, HEADER.size + cues * SECTION.size + 16, track_offset)
    (directory / "overlapping-sections.jydm").write_bytes(repaired(overlap))

    unsupported = bytearray(valid)
    tracks = section_index(unsupported, b"TRAK")
    struct.pack_into("<H", unsupported, HEADER.size + tracks * SECTION.size + 4, 2)
    (directory / "unsupported-version.jydm").write_bytes(repaired(unsupported))

    traversal = bytearray(valid)
    old_path = b"safe/song.aac\0"
    new_path = b"../x/song.aac\0"
    assert len(old_path) == len(new_path)
    location = traversal.index(old_path)
    traversal[location:location + len(old_path)] = new_path
    (directory / "traversal-path.jydm").write_bytes(repaired(traversal))

    optional = bytearray(valid)
    _, version, endian, header_size, directory_entry_size, section_count, track_count, flags, file_size, directory_offset, crc, reserved = HEADER.unpack_from(optional)
    insert_at = directory_offset + section_count * directory_entry_size
    optional[insert_at:insert_at] = SECTION.pack(b"XOPT", 99, 1, 0, 0, insert_at + SECTION.size)
    for index in range(section_count):
        entry = directory_offset + index * directory_entry_size
        struct.pack_into("<Q", optional, entry + 16, struct.unpack_from("<Q", optional, entry + 16)[0] + SECTION.size)
    struct.pack_into("<I", optional, 16, section_count + 1)
    struct.pack_into("<Q", optional, 28, file_size + SECTION.size)
    (directory / "unknown-optional.jydm").write_bytes(repaired(optional))


def main() -> int:
    compiler = shutil.which("c++")
    if not compiler:
        raise SystemExit("c++ compiler is required")
    with tempfile.TemporaryDirectory() as temporary:
        directory = Path(temporary)
        fixtures(directory)
        executable = directory / "firmware-reader-self-check"
        subprocess.run([
            compiler, "-std=c++11", "-Wall", "-Wextra", "-Werror",
            "-I", str(HOST_INCLUDE), "-I", str(ROOT / "src" / "Metadata"),
            str(READER), str(CHECK), "-o", str(executable),
        ], check=True)
        subprocess.run([str(executable), str(directory)], check=True)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/jayd-library/firmware_reader_self_check.cpp
#include "JaydMetadata.h"

#include <FS.h>
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <string>

using JaydMetadata::Cue;
using JaydMetadata::Grid;
using JaydMetadata::Phrase;
using JaydMetadata::Reader;
using JaydMetadata::Status;
using JaydMetadata::Track;

static std::string fixture(const char* directory, const char* name){
	return std::string(directory) + "/" + name;
}

static Status open(Reader& reader, const char* directory, const char* name){
	const std::string path = fixture(directory, name);
	return reader.open(fs::File(path.c_str()));
}

int main(int argc, char** argv){
	assert(argc == 2);
	Reader reader;
	Track track;

	assert(open(reader, argv[1], "missing.jydm") == Status::Missing);
	assert(open(reader, argv[1], "valid.jydm") == Status::Ready);
	assert(reader.trackCount() == 1);
	assert(reader.trackByIndex(0, track) == Status::Ready);
	assert(track.sampleRate == 44100);
	assert(track.durationFrames == 110250);
	assert(track.bpmMilli == 128125);
	assert(track.cueCount == 2);
	assert(track.gridCount == 1);
	assert(track.phraseCount == 1);

	char text[32];
	assert(reader.readString(track.provenance, text, sizeof(text)));
	assert(strcmp(text, "firmware-reader-self-check") == 0);
	uint32_t provenanceHash = 0;
	assert(reader.readStringHash(track.provenance, provenanceHash));
	assert(provenanceHash != 0);
	assert(reader.trackByPath("safe/song.aac", track) == Status::Ready);
	assert(reader.trackByPath(
		"safe/song.aac",
		track,
		track.fingerprint,
		track.sourceId
	) == Status::Ready);
	uint8_t staleFingerprint[16];
	memcpy(staleFingerprint, track.fingerprint, sizeof(staleFingerprint));
	staleFingerprint[0] ^= 1;
	assert(reader.trackByPath("safe/song.aac", track, staleFingerprint) == Status::Stale);
	uint8_t staleSourceId[16];
	memcpy(staleSourceId, track.sourceId, sizeof(staleSourceId));
	staleSourceId[0] ^= 1;
	assert(reader.trackByPath("safe/song.aac", track, nullptr, staleSourceId) == Status::Stale);
	assert(reader.trackByFingerprint(track.fingerprint, track) == Status::Ready);
	assert(reader.trackBySourceId(track.sourceId, track) == Status::Ready);
	assert(reader.trackByPath("missing.aac", track) == Status::Missing);

	Cue cue;
	Grid grid;
	Phrase phrase;
	assert(reader.readCue(track, 0, cue));
	assert(cue.positionFrames == 22050);
	assert(reader.readCue(track, 1, cue));
	assert(cue.kind == 2 && cue.lengthFrames == 22050);
	assert(reader.readGrid(track, 0, grid));
	assert(grid.beatNumber == 1 && grid.confidence == 9000);
	assert(reader.readPhrase(track, 0, phrase));
	assert(phrase.confidence == 8000);

	assert(open(reader, argv[1], "fractional-valid.jydm") == Status::Ready);
	assert(reader.trackByIndex(0, track) == Status::Ready);
	assert(reader.readCue(track, 0, cue));
	assert(cue.lengthFrames == Reader::NoFrame && cue.lengthNumerator == 3 &&
		cue.lengthDenominator == 2);
	assert(reader.readCue(track, 1, cue));
	assert(cue.positionFrames == Reader::NoFrame && cue.positionNumerator == 5 &&
		cue.positionDenominator == 2);
	assert(open(reader, argv[1], "fractional-boundary.jydm") == Status::Ready);
	assert(open(reader, argv[1], "fractional-denominator-zero.jydm") == Status::Corrupt);
	assert(open(reader, argv[1], "fractional-cue-over.jydm") == Status::Corrupt);
	assert(open(reader, argv[1], "fractional-loop-over.jydm") == Status::Corrupt);
	assert(open(reader, argv[1], "fractional-grid-over.jydm") == Status::Corrupt);
	assert(open(reader, argv[1], "fractional-phrase-over.jydm") == Status::Corrupt);
	assert(open(reader, argv[1], "fractional-64-bit-product-overflow.jydm") == Status::Corrupt);

	assert(open(reader, argv[1], "unknown-optional.jydm") == Status::Ready);
	assert(open(reader, argv[1], "corrupt-crc.jydm") == Status::Corrupt);
	assert(open(reader, argv[1], "truncated.jydm") == Status::Corrupt);
	assert(open(reader, argv[1], "oversized-count.jydm") == Status::Corrupt);
	assert(open(reader, argv[1], "offset-overflow.jydm") == Status::Corrupt);
	assert(open(reader, argv[1], "overlapping-sections.jydm") == Status::Corrupt);
	assert(open(reader, argv[1], "unsupported-version.jydm") == Status::Unsupported);
	assert(open(reader, argv[1], "traversal-path.jydm") == Status::Corrupt);
	assert(open(reader, argv[1], "duplicate-path.jydm") == Status::Ready);
	assert(reader.trackByPath("safe/song.aac", track) == Status::Stale);

	puts("firmware metadata reader self-check passed");
	return 0;
}
